snapdtool/fips_linux.go

```diff
@@ -0,0 +1,146 @@
+		{"usr/lib/riscv64-linux-gnu/libcrypto.so.3", "usr/lib/riscv64-linux-gnu/ossl-modules-3/fips.so"},
+		{"usr/lib/s390x-linux-gnu/libcrypto.so.3", "usr/lib/s390x-linux-gnu/ossl-modules-3/fips.so"},
+	}
+
+	for _, bundle := range fipsLibAndModulePathInSnapdSnap {
+		lib := filepath.Join(snapRoot, bundle.opensslLib)
+		module := filepath.Join(snapRoot, bundle.moduleFile)
+		if osutil.FileExists(lib) && osutil.FileExists(module) {
+			return lib, module
+		}
+	}
+	return "", ""
+}
+
+// MaybeSetupFIPS checks whether a system-wide FIPS mode is enabled and if
+// so sets up an environment such that the current process is able to use
+// libraries that are required for FIPS compliance and reexecs.
+func MaybeSetupFIPS() error {
+	enabled, err := fips.IsEnabled()
+	if err != nil {
+		return fmt.Errorf("cannot obtain FIPS status: %w", err)
+	}
+
+	if !enabled {
+		// FIPS not enabled do nothing
+		return nil
+	}
+
+	logger.Debugf("FIPS mode enabled system wide")
+
+	if os.Getenv("SNAPD_FIPS_BOOTSTRAP_DONE") == "1" {
+		// we've already been reexeced into FIPS mode and bootstrap was
+		// performed
+		logger.Debugf("FIPS bootstrap complete")
+
+		// if we reached this place, then the initialization was
+		// completed successfully and we can drop the environment
+		// variables, other processes which may be invoked by snapd will
+		// perform the initialization cycle on their own when needed
+		os.Unsetenv("GOFIPS")
+		os.Unsetenv("SNAPD_FIPS_BOOTSTRAP_DONE")
+		os.Unsetenv("OPENSSL_MODULES")
+		os.Unsetenv("GO_OPENSSL_VERSION_OVERRIDE")
+		return nil
+	}
+
+	snapdRev, err := osReadlink(snapdSnap)
+	if err != nil {
+		return err
+	}
+
+	currentRevSnapdSnap := filepath.Join(dirs.SnapMountDir, "snapd", snapdRev)
+
+	logger.Debugf("snapd snap: %s", currentRevSnapdSnap)
+
+	exe, err := osReadlink(selfExe)
+	if err != nil {
+		return err
+	}
+
+	logger.Debugf("self exe: %s", exe)
+
+	// on a classic system we need to be reexecuted from the snapd snap for
+	// the FIPS setup to be relevant, but on core we are not reexeced but
+	// running directly from the mount of the snapd snap under
+	// /usr/lib/snapd, yet we still want to set up the right environment
+	if release.OnClassic {
+		if !strings.HasPrefix(exe, currentRevSnapdSnap+"/") {
+			// this is only supported for reexecing from the snapd snap
+			return nil
+		}
+	}
+
+	lib, mod := findFIPSLibsAndModules(currentRevSnapdSnap)
+
+	env := append(os.Environ(), []string{
+		"SNAPD_FIPS_BOOTSTRAP_DONE=1",
+		// make FIPS mod required at runtime, if the module was not
+		// found or the setup is incorrect snapd will fail in a
+		// predictable way
+		"GOFIPS=1",
+	}...)
+
+	if mod != "" {
+		// version override uses the version suffix right after *.so.
+		libVer := strings.TrimPrefix(filepath.Ext(lib), ".")
+		logger.Debugf("found FIPS library and module at %s (ver %s) and %s", lib, libVer, mod)
+		// FIPS module not found, we still let the process reexec
```

Comment:
hmm perhaps I am missing something but this comment doesn't seem to match the code - we have just done a debug print saying we found the FIPS library and module but this comment says FIPS module not found...?

Reply:
thanks for spotting, I forgot to drop the comment after some earlier tweaks I made